src/common/RefCountedObj.h

```diff
@@ -76,6 +79,64 @@ struct RefCountedObject {
   CephContext *cct;
 };
 
+template <class RefCountedObjectType>
```

Comment:
i'm inclined to avoid using this machinery. it's complicated, and it forces its user to include the header defining the Message even it does not need to access the definition of the Message's type. instead of using member function, i think we should encourage the use of free function. put in other way, i think it would be more flexible to compose these building blocks instead of trying to use inheritance for reusing.

both standard library and boost already offer cast function of `std::const_pointer_cast<>` and `boost::const_pointer_cast<>` respectively. once we have the aliases and a factory function defined. they will just work fine altogether. the factory function can be a friend of the Message classes.

if we want to cast an `intrusive_ptr<Message> m` to `intrusive_ptr<const MDiscover>`, we can just put
```c++
 boost::static_pointer_cast<const MDiscover>(m)
```

it's pretty straightforward.

Reply:
Thanks for pointing this out Kefu.

> i'm inclined to avoid using this machinery. it's complicated, and it forces its user to include the header defining the Message even it does not need to access the definition of the Message's type.

Of course, there's no requirement to use the `ref_cast` machinery at all. It's just there as a convenience. In the general case, I think the message will already have been included as is done in the MDS at least.

In any case, I don't really want to spend more time on this set of PRs. I hope it's good enough and net improvement as-is for now. If someone wants to adjust this further, they have my blessing.

Reply:
As long as using the machinery is opt-in, I think this looks good!

Reply:
@liewegas @batrick is it acceptable if we replace 
```c++
MInodeFileCaps::msgref_cast(m)
```
with
```c++
ref_cast<MInodeFileCaps>(m)
```
?

if yes, i will try to add free functions for casting and creating messages. then remove the inheritance based machinery.

Reply:
As long as there is still a way to prevent `get`/`put` as in `RefCountedObjectSubTypeSafe` then I think it should be fine. I suspect you're also going to create a `ref_build<MInodeFileCaps>` (or similar) too?

Reply:
yeah, that's "functions for creating messages" for.

Reply:
Yes, sorry Kefu. Brain fart while reading!